# Using plink to for explotaroy analyses

This markdown describes the workflow from filtered VCF to GWAS for Plasmodium Knowlesi.

Helpful links:
PLINK: https://www.biostars.org/p/271694/
PLINK: https://plink.readthedocs.io/en/latest/plink_ex/
MOI: https://bahlolab.github.io/moimix/vignettes/introduction.html#estimating-moi-with-f_ws

## Required tools/libraries for Analyses

```{R,eval=F}
module load bcftools/1.12
module load java/jdk-8.40 
module load R/4.3.1 
export R_LIBS=/home/588/jw1542/.local/lib/R_4.3:$R_LIBS
export PATH=$PATH:/g/data/pq84/bin/plink2/
source /g/data/pq84/python-3.9.2/bin/activate

FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta"
VCF="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/04_Variant_calling/filtered/adjusted_filters/include_archived_samples"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"
REGIONS_TO_MASK="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/regions_to_mask.list"

cd /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS
```

## Exclude controls & MIT and API contigs

```{R,eval=F}
bcftools view $VCF/Consensus_filtered_SNPs.vcf.gz | head -n 100 | grep 'CHROM' | tr '\t' '\n' | grep -v 'ctrl' | tail -n+10 > filter_samples.txt

bcftools view $VCF/Consensus_filtered_SNPs.vcf.gz -r ordered_PKNH_01_v2,ordered_PKNH_02_v2,ordered_PKNH_03_v2,ordered_PKNH_04_v2,ordered_PKNH_05_v2,ordered_PKNH_06_v2,ordered_PKNH_07_v2,ordered_PKNH_08_v2,ordered_PKNH_09_v2,ordered_PKNH_10_v2,ordered_PKNH_11_v2,ordered_PKNH_12_v2,ordered_PKNH_13_v2,ordered_PKNH_14_v2 | \
  bcftools view -o Consensus_SNPs_subset.vcf.gz - -S filter_samples.txt
bcftools index -t Consensus_SNPs_subset.vcf.gz
```


## Calculate multiplicity of infection using moimix

To install MCMCpack on Gadi: https://opus.nci.org.au/display/Help/R 

Steps:
 - Mask problematic regions and subset to biallelic SNPs (assumption of moimix)
 - Convert VCF to GDS
 - Estimate BAF matrix 
    - estimate B-allele freq for each isolate. 
    - resulting bafmatrix stores estimates and genomic coordinates
 - Estimating MOI with binommix
    - estimated on per-isolate basis using binomial mixture model
    - to fit model - need a matrix of read counts and to set the value of k=2c where c is the number of clonal infections
 - Estimating MOI with Fws
    - alternative way of assessing clonality is to estmate Fws, with Fws < 0.95 indicating clonal infection (low MOI)


File(s):  
/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/moimix/moimix.pbs
/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/moimix/seqVCF2GDS.R

Version(s):
bcftools/1.12
R/4.1.0 

Submission: 
Single PBS script with nested R script.

Notes:

```{R, eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N moimix
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=60GB,ncpus=10
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load bcftools/1.12
module load R/4.1.0 
export R_LIBS_USER="/g/data/pq84/R"
cd /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS


echo "---------------------------------------"
echo "Update VCF"
bcftools index -t Consensus_SNPs_subset.vcf.gz
bcftools view Consensus_SNPs_subset.vcf.gz -M2 | bcftools view -f PASS - | bcftools view -e "MAC<2" -o moimix/SNP_subset_moimix.vcf.gz 

echo "---------------------------------------"
echo "Create GDS file"
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/moimix/seqVCF2GDS_2.R

echo "---------------------------------------"
echo "Finished"
```

```{R,eval=F}
# Converting VCF to GDS for estMOI
library(tidyverse)
library(SeqArray)
library(moimix)

seqVCF2GDS("moimix/Pf3D7_moimix_subset.vcf.gz", "moimix/Pf3D7_moimix_subset.gds", storage.option = "LZ4_RA")
isolates <- seqOpen("moimix/Pf3D7_moimix_subset_2.gds") # read into R
seqSummary(isolates)
sample.id <- seqGetData(isolates, "sample.id")

coords <- getCoordinates(isolates) # get genomic coordinates of all variants

## estimating BAF matrix 
isolate_baf <- bafMatrix(isolates)

baf_df <- isolate_baf$coords %>%
  cbind(
    as.data.frame(isolate_baf$baf_site) %>% 
      rename(baf = "isolate_baf$baf_site")
    ) %>%
  left_join(
    isolate_baf$baf_matrix %>%
      as.data.frame() %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("variant.id") %>%
      mutate(variant.id = as.numeric(variant.id)) 
    ) 

write_tsv(baf_df, "moimix/BAF_dataframe.tsv")


# Estimate Fws
set.seed(2023)

fws_all <- getFws(isolates) %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample") %>%
  rename("Proportion" = ".")

## Plot MOI distribution and export ######################

## output MOI data
fws_all %>%  
  write_tsv("moimix/fws_MOI.tsv")

## output samples with HIGH MOI data
fws_all %>%  
  filter(Proportion < 0.85) %>% 
  select(sample) %>%
  write_tsv("moimix/fws_high_MOI.tsv")
```

Plot Fws

```{R,eval=F}
library(tidyverse)
library(janitor)
library(viridis)

fws_all <- read_tsv("moimix/fws_MOI.tsv") %>%
  rename(Sample = sample) %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  mutate(Sample = str_replace(Sample, "_", "-")) 

metadata <- read_tsv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/full_metadata.tsv") %>% 
    mutate(sevemal = ifelse(grepl("UM", sevemal), "UM", .$sevemal)) %>% 
    mutate(sevemal = ifelse(sevemal == "1", "SM", ifelse(sevemal == "0", "UM", .$sevemal))) %>% 
    rename(Sample = studycode) %>% 
    mutate(district = str_to_title(district)) %>%
    add_column(Cluster = NA)

archived_data <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(district = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) %>% 
        rbind(read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        rename(district = Location) %>%
        add_column(Cluster = "Peninsular")
        ) %>%
        add_column(enroldate=NA, age = NA, sex=NA, ethnicity=NA, village=NA, occupation=NA, Hb=NA, platelets=NA,
        parasitemia=NA, fever_days=NA, respiratory_rate=NA, oxygen_saturation=NA, systolic_BP=NA, diastolic_BP=NA, G6PD=NA,
        pregnant=NA, creatinine=NA, study=NA, bicarbonate=NA, PCR=NA, sevemal=NA, bilirubin=NA, glucose=NA, bleeding_severity=NA)

metadata <- metadata %>%
    rbind(archived_data)

fws_meta <- fws_all %>% 
  left_join(metadata)

# Plot - column
fws_plot <- fws_meta %>%
  arrange(desc(Proportion)) %>%
  add_column(Sample_name = 1:nrow(.)) %>%
  mutate(MOI = ifelse(Proportion >= 0.98, "No MOI",
        ifelse(Proportion < 0.98 & Proportion > 0.85, "Low MOI", "High MOI"))) %>%
  ggplot(aes(x = Sample_name, y = Proportion, fill = MOI)) +
  geom_col() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  scale_fill_viridis_d()

ggsave("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS/moimix/fws_col_plot.png", width = 14, dpi = 300, fws_plot)

# Plot - box
fws_plot <- fws_meta %>% 
  ggplot(aes(x = district, y = Proportion, colour = district)) +
  geom_boxplot() +
  coord_flip() +
  scale_colour_viridis_d() +
  theme(legend.position = "none") + 
  ylab("District")
  
ggsave("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS/moimix/fws_box_plot_districts.png", dpi = 300, fws_plot)

# Plot - dot and line
fws_plot <- fws_meta %>%
  arrange(desc(Proportion)) %>%
  add_column(Sample_name = 1:nrow(.)) %>% 
  ggplot(aes(x = Sample_name, y = Proportion, colour = district), group = distict) +
  geom_point() +
  geom_line() +
  scale_colour_viridis_d() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  ylab("Proportion") +
  xlab("Samples")
  
ggsave("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS/moimix/fws_dot_line.png", dpi = 300, width = 12, fws_plot)

# Plot - density
fws_plot <- fws_meta %>%
  arrange(desc(Proportion)) %>%
  add_column(Sample_name = 1:nrow(.)) %>% 
  ggplot(aes(x = Proportion, colour = district), group = distict) +
  geom_density() +
  scale_colour_viridis_d(name = "District") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  ylab("Density")
  
ggsave("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS/moimix/fws_density.png", dpi = 300, width = 12, fws_plot)

kruskal.test(Proportion ~ district, data = fws_meta)

pairwise.wilcox.test(fws_meta$Proportion, g = fws_meta$district, p.adjust.method = "BH")$p.value %>% 
  as.data.frame() %>% 
  rownames_to_column("district_1") %>%
  pivot_longer(2:ncol(.), names_to = "district_2", values_to = "adj_p") %>%
  na.omit() %>% 
  write.table("moimix/wilcoxin_rank_sum_test.txt")
```

# Plot NRAF

```{R,eval=F}
library(tidyverse)

baf_df <- read_tsv("moimix/BAF_dataframe.tsv") %>% # Generated with previous script
  mutate(chromosome = str_remove(chromosome, "ordered_PKNH_0")) %>% 
  mutate(chromosome = str_remove(chromosome, "_v2")) %>%
  rename(REM111_2 = REM111, RK024_2 = RK024) %>%
  rename_with(~str_remove(., "_DK.*")) 

## Plot BAF for different Fws samples with ggplot
plot_baf <- function(DATA, SAMPLE){
DATA %>%
    ggplot(aes_string(x = "variant.id", y = SAMPLE, colour = "chromosome")) + # use aes string so that we can paste the str in from the function arguments - passing the sample in not as a string does not work
    geom_point() +
    scale_colour_manual(values = c(rep_len(c("#404788FF", "#1F968BFF"), length(unique(baf_df$chromosome))))) +
    xlab("Chromosome") +
    ylab("NRAF") + 
    theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = "bottom", 
        legend.title = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
    guides(colour = guide_legend(nrow = 1))
}
# top 20
sample_list <- c("PK_SB_DNA_083", "ERR985395", "QEM662", "SK004", "QEM689", "NQE2_79", "SK028", "SK029P4", "QEM545", "NKM1_39", "PK_SB_DNA_006", "QEM306", "QEM337", "ERR985405", "REM116", "PCM218", "QDM007", "KK055", "TK022", "NKT3_15")

for (i in sample_list){
  baf_plot <- plot_baf(baf_df, paste0(i))
  ggsave(paste0("moimix/",i,"_NRAF.png"), width = 14, dpi = 300, baf_plot)
}
```


## Exclude samples with high MOI

File(s):  
/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/Exclude_high_fws_samples.pbs

Version(s):
bcftools/1.12

Submission: 
Single PBS script.

Notes:
To get list of samples:
cat moimix/fws_high_MOI.tsv | tr '\n' ',' 

```{R,eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N Exclude_high_fws
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=32GB,ncpus=4
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load bcftools/1.12
cd /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS

echo "---------------------------------------"
echo "Exclude high fws samples"
bcftools view Consensus_SNPs_subset.vcf.gz -s ^KK046_DKDN230007989-1A_H2HTMDSX7_L2,KK048_DKDN230007688-1A_H2HG5DSX7_L4,KK055_DKDN230008054-1A_H2H7YDSX7_L3,KK069_DKDN230008059-1A_H2H7YDSX7_L3,KK118_DKDN230008074-1A_H2H7YDSX7_L3,KK119_DKDN230007935-1A_H2HTMDSX7_L1,KK123,LK003_DKDN230008083-1A_H2H7YDSX7_L3,LK035_DKDN230013624-1A_H5YHNDSX7_L3,MK022_DKDN230007678-1A_H2HG5DSX7_L4,MK026_DKDN230007684-1A_H2HG5DSX7_L4,MK043_DKDN230013629-1A_H5YHNDSX7_L3,NKB3_57_DKDN230007610-1A_H2H7YDSX7_L3,NKM1_39_DKDN230007715-1A_H2H7YDSX7_L3,NKM4_130_DKDN230007754-1A_H2H7YDSX7_L2,NKT3_15_DKDN230007606-1A_H2HG5DSX7_L4,NKU2_118_DKDN230007729-1A_H5YHNDSX7_L3,NKU3_52_DKDN230013632-1A_H5YHNDSX7_L3,NQE2_152_DKDN230007745-1A_H2H7YDSX7_L2,NQE2_79,NQE4_103_DKDN230007630-1A_H2HG5DSX7_L4,NRU4_53_DKDN230007636-1A_H2HG5DSX7_L4,NRU4_82_DKDN230007670-1A_H2H7YDSX7_L2,NSD31_DKDN220017460-1A_H7YMHDSX5_L4,NSD3_16_DKDN230008118-1A_H2HTMDSX7_L1,NTW4_99_DKDN230007753-1A_H2H7YDSX7_L2,PCM077_DKDN230007882-1A_H2HTMDSX7_L2,PCM100_DKDN230007875-1A_H2HTMDSX7_L1,PCM126_DKDN230007780-1A_H2H7YDSX7_L2,PCM143_DKDN230007893-1A_H2HTMDSX7_L2,PCM186_DKDN220017456-1A_H7YMHDSX5_L4,PCM215_DKDN230007786-1A_H2H7YDSX7_L2,PCM218_DKDN230007787-1A_H2H7YDSX7_L2,PCM224_DKDN230007789-1A_H2HTMDSX7_L1,PCM235_DKDN230007792-1A_H2H7YDSX7_L2,PCM255_DKDN230007799-1A_H2HTMDSX7_L1,PCM259_DKDN230007858-1A_H2HTMDSX7_L2,PCM280,PCM286_DKDN220017434-1A_H7YMHDSX5_L4,PCM309_DKDN220017421-1A_H7YMHDSX5_L4,PCM319_DKDN230007811-1A_H2H7YDSX7_L2,PCM320_DKDN230007709-1A_H2HG5DSX7_L4,PCM322_DKDN230007768-1A_H2H7YDSX7_L2,PCM331_DKDN230007869-1A_H2HTMDSX7_L2,PCM332_DKDN220017445-1A_H7YMHDSX5_L4,PCM377_DKDN230013646-1A_H5YHNDSX7_L3,PCM379_DKDN230007862-1A_H2HTMDSX7_L2,PCM389_DKDN230007865-1A_H2HTMDSX7_L2,PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4,PK_SB_DNA_032_DKDL210002161-1a_HWHGKDSXY_L4,PK_SB_DNA_039_DKDL210002168-1a_HWHGKDSXY_L4,PK_SB_DNA_048_DKDL210002177-1a_HWHGKDSXY_L4,PK_SB_DNA_051_DKDL210002180-1a_HWHGKDSXY_L4,PK_SB_DNA_055_DKDL210002184-1a_HWHGKDSXY_L4,PK_SB_DNA_083_DKDL210002212-1a_HWHGKDSXY_L4,QDM007_DKDN220017469-1A_H7YMHDSX5_L4,QDM092,QDM136_DKDN230007928-1A_H2HTMDSX7_L1,QDM1408_DKDN230007836-1A_H2HTMDSX7_L2,QDM201_DKDN230007980-1A_H2HTMDSX7_L2,QDM492_DKDN230007819-1A_H2H7YDSX7_L2,QEM047_DKDN230008000-1A_H2HTMDSX7_L2,QEM215_DKDN230008057-1A_H2H7YDSX7_L3,QEM306_DKDN230008014-1A_H5YHNDSX7_L3,QEM337_DKDN230008007-1A_H5YHNDSX7_L3,QEM446_DKDN230008021-1A_H5YHNDSX7_L3,QEM469_DKDN230007907-1A_H2HTMDSX7_L2,QEM542_DKDN230007901-1A_H2HTMDSX7_L2,QEM545_DKDN230007903-1A_H2HTMDSX7_L1,QEM662,QEM673,QEM688_DKDN230007990-1A_H5YHNDSX7_L3,QEM689_DKDN230007992-1A_H5YHNDSX7_L3,QEM706_DKDN230007687-1A_H2HG5DSX7_L4,QEM929_DKDN230007929-1A_H2HTMDSX7_L2,REM045_DKDN230007702-1A_H2HG5DSX7_L4,REM116_DKDN230007704-1A_H2HG5DSX7_L4,RK013_DKDN230007691-1A_H2H7YDSX7_L2,RK027_DKDN220017425-1A_H7YMHDSX5_L4,RK040_DKDN230013630-1A_H5YHNDSX7_L3,SK004_DKDN220017483-1A_H7YMHDSX5_L4,SK028_DKDN220017487-1A_H7YMHDSX5_L4,SK029P4_DKDN230007934-1A_H5YHNDSX7_L3,SK031_DKDN220017488-1A_H7YMHDSX5_L4,TF003_DKDN230013612-1A_H5YHNDSX7_L3,TK022_DKDN220017479-1A_H7YMHDSX5_L4 \
  | bcftools view -f PASS - | bcftools view -e "MAC<2" -o Consensus_SNPs_subset_no_MOI.vcf.gz

echo "---------------------------------------"
echo "Finished"
```

## Normalise VCF & convert filtered vcf to bcf

File(s):  
/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/vcf_to_bcf.pbs

Version(s):
bcftools/1.12

Submission: 
Single PBS script.

Notes:
 - Mask hypervariable regions with bcftools (already done in previous step)
 - Ensure that multi-allelic calls are split (1st part)
 - Left-align indels (2nd part)
 - Set the ID field to a unique value: CHROM:POS:REF:ALT (3rd part)
 
 --indiv-sort - specify how samples will be sorted - use to be consistent when adding metadata downstream 

```{R,eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N vcf_to_bcf
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=16GB,ncpus=2
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load bcftools/1.12
FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta"
cd /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/outputs/05_Analyses/GWAS

echo "---------------------------------------"
echo "VCF to BCF"
bcftools view Consensus_SNPs_subset_no_MOI.vcf.gz | \
  sed 's/AD,Number=R/AD,Number=./' | \
  bcftools norm -m-any | \
  bcftools norm --check-ref w -f $FASTA/strain_A1_H.1.Icor.fasta | \
  bcftools annotate -Ob -x ID -I +'%CHROM:%POS:%REF:%ALT:%ID' \
  > Consensus_SNPs_subset_no_MOI.bcf

echo "---------------------------------------"
echo "FINISHED"
```


## Convert to PLINK format 

Generic PLINK function to produce generic PLINK files
First create sample_order file & regions_to_mask file 

The regions we masked are: https://plasmodb.org/plasmo/service/users/current/steps/429386433/reports/attributesTabular & https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5918592/ and have to follow this format: https://www.cog-genomics.org/plink/1.9/input#make_set.
To find other hypervariable regions that may need to be masked, you can plot data from Pk.bim (created in step below) to acquire SNP frequencies/density in a sliding window approach.

Need to reset chromosomes for fastLMM downstream.

```{bash,eval=F}
bcftools view Consensus_SNPs_subset_no_MOI.bcf| sed -n '/#CHROM/,$p' | head -n 1 | tr '\t' '\n' |  tail -n+10 | awk '{print $1 "\t" $1}' > sample_order.txt

awk '{sub(/ordered_PKNH_/,"",$1)}1' Pk.bim | awk '{sub(/_v2/,"",$1)}1' | awk '{print $2 "\t" $1}' > chrom_update.txt

plink --bcf Consensus_SNPs_subset_no_MOI.bcf \
  --keep-allele-order \
  --vcf-idspace-to _ \
  --double-id \
  --allow-extra-chr 0 \
  --update-chr chrom_update.txt \
  --make-bed \
  --indiv-sort file sample_order.txt \
  --exclude 'range' regions_to_mask.txt \
  --allow-no-sex \
  --out Pk
```

## To remove duplicates

```{bash,eval=F}
cut -f 2 Pk.fam | uniq -d > Pk.dups
plink --bfile Pk --recode --remove Pk.dups --out Pk
```

### For SNP density across the genome

```{R,eval=f}
# Load packages
library(tidyverse)

# Read in data
snp_data <- read_table("Pk.bim", col_names = F) %>%
    select(X2) %>%
    separate(X2, into = c("Contig", "Pos", "Major", "Minor", "Nothing"), sep = ":") %>%
    mutate(Pos = as.numeric(Pos)) %>%
    select(1:2) %>%
    arrange(Contig, Pos)

# Create window 
window_size <- 10000

snp_plot <- snp_data %>%
    mutate(Window = (floor(Pos/window_size) * window_size)+ (window_size/2)) %>%
    group_by(Contig, Window) %>%
    summarise(SNPs = n()) %>%
    ggplot(aes(x = Window, y = SNPs, colour = Contig)) +
    geom_point() +
    facet_wrap(~Contig) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    xlab("Chromosome position (10K Windows)")

ggsave("SNP_density.png", dpi = 300, width = 14, snp_plot)

snp_plot <- snp_data %>%
    mutate(Window = (floor(Pos/window_size) * window_size)+ (window_size/2)) %>%
    group_by(Contig, Window) %>%
    summarise(SNPs = n()) %>%
    ggplot(aes(x = Window, y = SNPs, colour = Contig)) +
    geom_line() +
    facet_wrap(~Contig) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    xlab("Chromosome position (10K Windows)")

ggsave("SNP_density_line.png", dpi = 300, width = 14, snp_plot)   
```


## Add metadata to PLINK fam

```{R,eval=F}
# Packages
library(tidyverse)
library(janitor)
library(data.table)

# Metadata
metadata <- read_csv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/full_dataset_011223.csv") %>%
    rename(Sample = studycode) %>% 
    add_column(Cluster = NA) %>% 
    mutate(sevemal = ifelse(grepl("Um", sevemal), "Um", 
        ifelse(sevemal == "1", "Sm",
        ifelse(sevemal == "0", "Um", .$sevemal)))) %>%
    mutate(Sample = str_to_upper(Sample))

original_metadata <- readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>%
    select(1:3) %>%
    mutate(subjectid = ifelse(str_length(subjectid) == 1, paste0("00", subjectid), # if subjectid length = 1 paste 00
        ifelse(str_length(subjectid) == 2, paste0("0", subjectid), # if not, then if subjectid length = 2 paste 0
        .$subjectid))) %>%
    unite(Sample2, c("group", "subjectid"), sep ="") %>% 
    rename(Sample = sampleid)

archived_data <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(district = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) %>% 
        rbind(read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        rename(district = Location) %>%
        add_column(Cluster = "Peninsular")
        ) %>%
        add_column(enroldate=NA, age = NA, sex=NA, ethnicity=NA, village=NA, occupation=NA, Hb=NA, platelets=NA,
        parasitemia=NA, fever=NA, respiratory_rate=NA, oxygen_saturation=NA, systolic_BP=NA, diastolic_BP=NA, G6PD=NA,
        pregnant=NA, creatinine=NA, study=NA, bicarbonate=NA, PCR=NA, sevemal=NA, bilirubin=NA, glucose=NA, bleeding_severity=NA)

# To check colnames are the same
colnames(metadata) %>% 
    as.data.frame() %>% 
    rename(V1 = ".") %>% 
    arrange(V1) %>% 
    cbind(colnames(archived_data) %>% 
        as.data.frame() %>% 
        rename(V2 = ".") %>% 
        arrange(V2 )
    )

metadata <- metadata %>%
    rbind(archived_data) %>%
    select(Sample, sevemal)

Plink_fam <- read.table("Pk.fam", sep =" ") %>% 
    filter(!grepl("PK_SB_DNA", V1)) %>% # original batch - name does not line up with metadata
    mutate(Sample = str_remove(V1, "_DK.*")) %>% 
    mutate(Sample = str_replace(Sample, "_", "-")) %>% 
    left_join(metadata) %>% 
    select(-Sample) %>% 
    rbind(
        read.table("Pk.fam", sep =" ") %>% 
            filter(grepl("PK_SB_DNA", V1)) %>%
            mutate(Sample = str_remove(V1, "_DK.*")) %>% 
            left_join(original_metadata) %>% 
            mutate(Sample = Sample2) %>% 
            left_join(metadata) %>%
            select(-c(Sample, Sample2))
    )

write.table(Plink_fam, "Pk.fam", quote = FALSE, sep = " ", col.names = FALSE, row.names = FALSE)

Plink_fam %>% 
  select(-c(2:6)) %>% 
  write_csv("Pk.csv")
```

## Calculate minor allele frequency, genotypic missingness

```{eval=F}
plink --bfile Pk \
  --freq \
  --missing \
  --allow-no-sex \
  --out Pk
```

 - N_MISS =Number of missing SNPs
 - N_GENO =Number of non-obligatory missing genotypes
 - F_MISS =Proportion of missing SNPs

## Plot distribution of MAF and genotypic missingness

```{bash,eval=F}
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/06_Analyses/maf_dist.R
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/06_Analyses/miss_dist.R
```

### maf_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)

# Read in data
Pk_maf <- read_table("Pk.frq", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*"))

# Density of MAF by SNP
Pk_maf_density <- Pk_maf  %>%
    ggplot(aes(x = MAF)) +
    geom_density(alpha=.3) 

ggsave("filtering/Pk_maf_density.png", dpi = 300, Pk_maf_density)

# Plot MAF by chr
Pk_maf_chr <- Pk_maf %>% 
    mutate(SNP = str_remove(SNP, "ordered_PKNH_"),
        SNP = str_remove(SNP,"_v2.*")) %>% 
    group_by(SNP) %>%
    summarise(MAF = mean(MAF)) %>%
    ggplot(aes(x = SNP, y = MAF)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45), legend.position = "none") +
    scale_fill_viridis_d() 

ggsave("filtering/Pk_maf_plot_chr.png", dpi = 300, Pk_maf_chr)

# Summary stats

Pk_maf %>% 
    mutate(SNP = str_remove(SNP, "ordered_PKNH_"),
        SNP = str_remove(SNP,"_v2.*")) %>% 
    group_by(SNP) %>%
    summarise(MAF_mean = mean(MAF),
            MAF_SD = sd(MAF),
            MAF_SE = (sd(MAF))/sqrt(n())) %>%
rbind(
    Pk_maf %>%
        summarise(MAF_mean = mean(MAF),
                MAF_SD = sd(MAF),
                MAF_SE = (sd(MAF))/sqrt(n())) %>%
        add_column(SNP = "Total")
    ) %>%
    write_csv(col_names = T, "filtering/Pk_maf_summary.csv")
```

### miss_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)
library(data.table)

# Sample-based missing data
## Read in data
sample_mis <- read_table("Pk.imiss", col_names=T) 

# Density of miss by sample
sample_miss_density <- sample_mis %>% 
    rename(Sample = IID,
        Geno_Miss = F_MISS) %>%
    select(2:ncol(.)) %>%
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.05, colour = "#1F968BFF") 

ggsave("filtering/sample_miss_density_plot.png", width = 12, dpi = 600, sample_miss_density)

# Mean missingness
metadata <- read_tsv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/full_metadata.tsv") %>% 
    mutate(sevemal = ifelse(grepl("UM", sevemal), "UM", .$sevemal)) %>% 
    mutate(sevemal = ifelse(sevemal == "1", "SM", ifelse(sevemal == "0", "UM", .$sevemal))) %>% 
    rename(Sample = studycode) %>% 
    mutate(district = str_to_title(district)) %>%
    add_column(Cluster = NA)

archived_data <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(district = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) %>% 
        rbind(read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        rename(district = Location) %>%
        add_column(Cluster = "Peninsular")
        ) %>%
        add_column(enroldate=NA, age = NA, sex=NA, ethnicity=NA, village=NA, occupation=NA, Hb=NA, platelets=NA,
        parasitemia=NA, fever_days=NA, respiratory_rate=NA, oxygen_saturation=NA, systolic_BP=NA, diastolic_BP=NA, G6PD=NA,
        pregnant=NA, creatinine=NA, study=NA, bicarbonate=NA, PCR=NA, sevemal=NA, bilirubin=NA, glucose=NA, bleeding_severity=NA)

metadata <- metadata %>%
    rbind(archived_data)


sample_mis %>%
    left_join(
        metadata %>%
            rename(FID = Sample)
    ) %>%
    summarise(mean = mean(F_MISS), sd = sd(F_MISS), min = min(F_MISS), max = max(F_MISS)) %>% 
    write_csv("filtering/missingness.csv")


# Variant-based missing data
## Read in data
var_miss <- read_table("Pk.lmiss", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*")) %>%
    rename(Geno_Miss = F_MISS,
            Chr = CHR)

# Density of miss by variant
var_miss_density <- var_miss %>% 
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.2, colour = "#1F968BFF") +
    geom_vline(xintercept = 0.1, colour = "#1F968BFF") +
    geom_vline(xintercept = 0.05, colour = "#1F968BFF") 

ggsave("filtering/variant_miss_density_plot.png", width = 12, dpi = 600, var_miss_density)


# Summary stats

miss_summary <- var_miss %>%
    summarise(Miss_mean = mean(Geno_Miss),
                Miss_SD = sd(Geno_Miss),
                Miss_SE = (sd(Geno_Miss))/sqrt(n()))

write_csv(miss_summary, "filtering/Pk_miss_summary.csv")
```

## Apply filters for genotypic missingess

 - first recode creates the intial .map and .ped files
 - recode generates new .ped and .map files
 - --make-bed generates .bed, .bim and .fam files
 - -mind filters per sample (missingness)
 - -geno filters per variant (missingness)

**FILTERED:**
 - exclude variants with > 20% missing alleles
 - exclude variants with minor allele frequencies < 5%
 - exclude samples with > 5% missing alleles

Values selected based on the distrbutions plotted previously.

MAF (=Minor Allele Frequency) = P(geno=1, Aa)x0.5 + P(geno=2, aa)

```{bash,eval=F}
plink --bfile Pk --recode --set-hh-missing --out cleaned
plink --file cleaned --geno 0.25 --make-bed --out cleaned
plink --bfile cleaned --recode --out cleaned
plink --file cleaned --mind 0.25 --make-bed --out cleaned
plink --bfile cleaned --recode --out cleaned
plink --file cleaned --maf 0.05 --make-bed --out cleaned
```


#########################################################################################################

## Use PLINK to create PCA and MDS data, and a distance matrix (identity-by-state)

The different functions:
    --cluster is MDS
    --distance is identity by state
    --genome is identity by decent
        unbounded - "The underlying P(IBD=0/1/2) estimator sometimes yields numbers outside the range [0,1]; by default, these are clipped. The 'unbounded' modifier turns off this clipping."
        nudge - "Then, if PI_HAT2 < P(IBD=2), 'nudge' adjusts the final estimates to P(IBD=0) := (1-p2), P(IBD=1) := 2p(1-p), and P(IBD=2) := p2, where p is the current PI_HAT."

```{bash,eval=F}
plink --bfile cleaned --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --pca --out Pk
plink --bfile cleaned --distance square --out Pk
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/scripts/05_Analyses/GWAS/ordination_plots.R
```

### ordination_plots.R

```{R,eval=F}
#Load packages
library(tidyverse)
library(ape)
library(ggtree)
library(viridis)

# PCA

## Read in data
pca <- read_table("Pk.eigenvec", col_names=F) %>%
    select(-1) %>%
    rename("sampleid" = "X2") %>%
    rename_at(vars(starts_with("X")), ~str_replace(., "X.*", paste0("PC", seq_along(.))))

eigenval <- scan("Pk.eigenval")

## Add metadata 
metadata <- read_tsv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/full_metadata.tsv") %>% 
    mutate(sevemal = ifelse(grepl("UM", sevemal), "UM", .$sevemal)) %>% 
    mutate(sevemal = ifelse(sevemal == "1", "SM", ifelse(sevemal == "0", "UM", .$sevemal))) %>% 
    rename(Sample = studycode) %>% 
    mutate(district = str_to_title(district)) %>%
    add_column(Cluster = NA)

archived_data <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(district = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) %>% 
        rbind(read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        rename(district = Location) %>%
        add_column(Cluster = "Peninsular")
        ) %>%
        add_column(enroldate=NA, age = NA, sex=NA, ethnicity=NA, village=NA, occupation=NA, Hb=NA, platelets=NA,
        parasitemia=NA, fever_days=NA, respiratory_rate=NA, oxygen_saturation=NA, systolic_BP=NA, diastolic_BP=NA, G6PD=NA,
        pregnant=NA, creatinine=NA, study=NA, bicarbonate=NA, PCR=NA, sevemal=NA, bilirubin=NA, glucose=NA, bleeding_severity=NA)


metadata <- metadata %>%
    rbind(archived_data) %>%
  mutate_at(c("district", "village"), str_to_title) %>%
  mutate(Sample2 = Sample)

pca <- pca %>%
    rename(Sample = sampleid) %>%
    mutate(Sample2 = str_remove(Sample, "_DK.*")) %>% # we need the original names downstream for the NJT
    left_join(metadata, by = "Sample2") 


## Percentage variance explained by each PC
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)

pve_plot <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100) %>%
    ggplot(aes(PC, pve)) + 
    geom_bar(stat = "identity") + 
    ylab("Percentage variance explained") + 
    theme_light()

ggsave("clustering/Percentage_variance_explained.png", dpi=600, pve_plot)


## PCA - clusters
pca <- pca %>% 
    mutate(Cluster = ifelse(PC1 < -0.05 , "Mn", 
        ifelse(PC1 > 0 & PC2 < 0, "Mf", 
        ifelse(PC2 > 0.1, "Peninsular", .$Cluster)))) 

pca_plot <- ggplot(pca, aes(PC1, PC2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF")) +
    coord_equal() + 
    theme_light() + 
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + 
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))

ggsave("clustering/PCA_cluster.png", dpi = 600, pca_plot)

## PCA - districts
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)

pca_plot <- ggplot(pca, aes(PC1, PC2, colour = district)) + 
    geom_point(size = 3) +
    scale_color_viridis_d() +
    coord_equal() + 
    theme_light() + 
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + 
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))

ggsave("clustering/PCA_districts.png", dpi = 600, pca_plot)

# Updated metadata with clusters
metadata <- pca %>%
    select(Sample.x, enroldate:ncol(.)) %>%
    rename(Sample = Sample.x)

# Summarise sample clusters
metadata  %>% 
    group_by(Cluster) %>%
    summarise(n = n()) %>% 
    write_tsv("clustering/Sample_cluster_sumamry.tsv")



############################################################################################################################################################

# MDS - clusters

## Read in data
mds <- read_table("Pk.mds", col_names=T) %>%
    select(-c("FID", "X14")) %>%
    rename("sampleid" = "IID") %>%
    rename_at(vars(starts_with("C")), ~str_replace(., "C", "MDS"))

## Add metadata 
mds <- mds %>%
    rename(Sample = sampleid) %>% 
    left_join(metadata)

## Plot MDS
mds_plot <- ggplot(mds, aes(MDS1, MDS2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_viridis_d() +
    coord_equal() + 
    theme_light()

ggsave("clustering/MDS_cluster.png", dpi=600, mds_plot)

############################################################################################################################################################


# Neighbour-joining tree

## Create distance matrix from PLINK data and build NJT
NJT_ID <- read_table("Pk.dist.id", col_names=F) %>%
    as.data.frame()

NJT_matrix <- read_table("Pk.dist", col_names=NJT_ID$X1) %>%
    as.data.frame() %>%
    add_column(Row_Names = NJT_ID$X1) %>%
    column_to_rownames("Row_Names") %>%
    as.matrix()

NJT_tree <- nj(NJT_matrix)

## Plot tree


#Plot 
options(ignore.negative.edge=TRUE)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF"))

ggsave("clustering/NJT_tree_Cluster_unrooted.png", dpi = 300, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = district)) %<+% metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) 

ggsave("clustering/NJT_tree_district_unrooted.png", dpi = 300, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank())  +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF")) +
    geom_tiplab()

ggsave("clustering/NJT_tree_Cluster_unrooted_labelled.png", dpi = 300, height =30, width = 30, NJT_tree_plot)
```

